[RHCLOUD-19378] Feature migration fix duplicated tenants subresources

MikelAlejoBR opened the pull request:
This migration grabs all the duplicated tenants —the ones that have the
same "external_tenant" or "org_id" column values— and picks the first
one as the main one.

Then, for the other duplicates of this tenant, it updates all the
subresources' —application authentications, applications, authentications,
endpoints, sources and source rhc connections— tenant id to the main tenant.

Finally, it updates the duplicated tenants with an easily identifiable
value for the "external_tenant" and "org_id" columns.

## Links

[[RHCLOUD-19378]](https://issues.redhat.com/browse/RHCLOUD-19378)

MikelAlejoBR committed:
feature: migration for updating the duplicated tenants' subresources

This migration grabs all the duplicated tenants —the ones that have the
same "external_tenant" or "org_id" column values— and picks the first
one as the main one.

Then, for the other duplicates of this tenant, it updates all the
subresources' —application authentications, applications, authentications,
endpoints, sources and source rhc connections— tenant id to the main tenant.

Finally, it updates the duplicated tenants with an easily identifiable
value for the "external_tenant" and "org_id" columns.

petracihalova reviewed (approved):
nice job !!!

lpichler commented on db/migrations/20220608090500_remove_duplicated_tenant_ids_org_ids.go:
```diff
@@ -0,0 +1,255 @@
+						HAVING
+							count("texten"."external_tenant") > 1
+						AND
+							"texten"."external_tenant" != ''
```
do we also need to cover when `external_tenant` is not `null` ?

MikelAlejoBR replied:
Actually I think we can remove the `"texten"."external_tenant" != ''`, since the previous migration takes care of doing the `'' -> NULL` conversion on the database.

As to your question: no, we don't need that condition. The `AND "external_tenant" != ''` was there because the empty string counted as a "duplicate" as well. So, if we had the following:



id | external_tenant | org_id
-- | -- | --
1 | 12345 | 9876543210
2 |   | abcde
3 |   | deeehg

It would pick up rows with id `2` and `3` as if they were a duplicate, when in reality the only duplicate there is the empty string on the `external_tenant` column. With `NULL`s that doesn't happen because `NULL == NULL -> false`. You can test this yourself:

```sql
INSERT INTO
	"tenants"("external_tenant", "org_id", "created_at", "updated_at")
VALUES
	(NULL, 'abcde', now(), now()),
	(NULL, 'edcba', now(), now()),
	('', 'hello', now(), now()),
	('', 'notsame', now(), now())
```

And then run the query without the `AND` condition:


```sql
SELECT
	"t"."id", "t"."name", "t"."description", "t"."external_tenant", "t"."org_id", "t"."created_at", "t"."updated_at"
FROM
	"tenants" AS "t"
WHERE
	"t"."external_tenant" IN (
		SELECT
			"texten"."external_tenant"
		FROM
			"tenants" AS "texten"
		GROUP BY
			"texten"."external_tenant"
		HAVING
			count("texten"."external_tenant") > 1
	)
OR
	"t"."org_id" IN (
		SELECT
			"torgid"."org_id"
		FROM
			"tenants" AS "torgid"
		GROUP BY
			"torgid"."org_id"
		HAVING
			count("torgid"."org_id") > 1
	)
```

You'll see that the query picks up the two rows with the empty string, when it shouldn't. However, as I said, in the previous migration I take care of the empty strings on both columns, and also I ran a query on AppInterface on the stage and prod environments to check if we have empty strings on either of the columns and we don't.

lpichler replied:
thanks for detailed explanation, now I understand

MikelAlejoBR replied:
Here are the results of the queries that I ran on both clusters: https://issues.redhat.com/browse/RHCLOUD-19802 (restricted to the team for confidentiality reasons). The MR that I sent to AppInterface checked for `external_tenant = ''`, `"external_tenant" = '' AND "org_id" != ''` and `"external_tenant" = '' AND "org_id" = ''` and yielded zero results, so we should be safe.

lpichler commented on db/migrations/20220608090500_remove_duplicated_tenant_ids_org_ids.go:
```diff
@@ -0,0 +1,255 @@
+
+				// Group all "external_tenant" and "org_id"s into the above defined maps.
+				for _, t := range tenants {
+					// Append the ID to the external tenants map.
```
maybe log message here ? to see which tenants are duplicate

MikelAlejoBR replied:
Thank you :)

lpichler commented on db/migrations/20220608090500_remove_duplicated_tenant_ids_org_ids.go:
```diff
@@ -0,0 +1,255 @@
+
+				// Start processing the external tenants and grouping them into the main tenant.
+				for externalTenant, ids := range externalTenantIds {
+					// Grab the first tenant ID as the main one.
```
maybe log message "..processing externalTenant..."

MikelAlejoBR replied:
Included, thanks!

lpichler commented on db/migrations/20220608090500_remove_duplicated_tenant_ids_org_ids.go:
```diff
@@ -0,0 +1,255 @@
+				// Finish by processing the org IDs to group them on the main tenant. The difference with the above
+				// code is that if an ID has already been processed, it will be skipped to avoid problems.
+				for orgId, ids := range orgIdIds {
+					// Grab the first tenant ID as the main one.
```
maybe log message "..processing orgId..."

MikelAlejoBR replied:
Done, thank you!

MikelAlejoBR committed:
refactor: remove "AND" condition on the "HAVING" condition

A previous migration takes care of removing the empty spaces from the
two columns that are being checked for them. Therefore, we can simply
remove the condition, since "NULL" values don't count towards
duplicates.

lpichler commented on db/migrations/20220608090500_remove_duplicated_tenant_ids_org_ids.go:
```diff
@@ -0,0 +1,251 @@
+			// This query selects all the tenants which have duplicated "external_tenant" or "org_id" column values.
+			duplicatesSql := `
+				SELECT
+					"t"."id", "t"."name", "t"."description", "t"."external_tenant", "t"."org_id", "t"."created_at", "t"."updated_at"
```
do we need all columns here ?

MikelAlejoBR replied:
You are right, we can probably narrow it down just to the "id", "external_tenant" and "org_id" columns

MikelAlejoBR replied:
Well now that I think about it, I log the entire tenant object with a `"%v#` on the logging statement. So maybe we can leave it like this, or simply a `*`?

lpichler replied:
yes we leave as it is, thanks !

MikelAlejoBR committed:
refactor: add more logs to the migration

Improves traceability and will help us debug things if something goes wrong.

syncrou commented on db/migrations/20220608090500_remove_duplicated_tenant_ids_org_ids.go:
```diff
@@ -0,0 +1,263 @@
+	return nil
+}
+
+// markExternalTenantAsDuplicate marks the old duplicated tenant as duplicate. It modifies the "org_id" column to add a
```
Small nit, but it looks like the documentation for `markOrgIdAsDuplicate` is actually the copied documenation for `markExternalTenantAsDuplicate`

syncrou replied:
Or the name of the function is - I see the `org_id` reference later on.

MikelAlejoBR replied:
Yes, initially I was doing one pass and updating both columns to the same value, but then I thought that doing a second pass with org ids wouldn't hurt, since the redundant updates would essentially be NOPs. That's why we have two functions for the migration. I'll update the docs.

syncrou reviewed (commented):
Very small documentation nit, but otherwise this looks great.

I have 1 question about the transaction and bulk updates:

Is there any possibility that the bulk transaction could fail based on running so many updates and hence kill the entire process ( ie maybe due to hitting some rate limit on RDS. )

If yes, then obviously no destructive behavior will happen, but the migration will not be able to ever complete.

If yes, then potentially we may need to move to an atomic update where an update is called per tenant id and that is in a transaction. If that transaction fails then all the other successful updates stay as is.  Then when the problem ( that killed the last transaction ) is fixed the migration can pick up where it left off instead of starting from scratch.

I realize there isn't an answer to this now, but thought I would bring it up since I was concerned about it.
